File: samples/C++/BarcodeReaderDemo/BarcodeReaderDemo.cpp
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#if defined(_WIN64) || defined(_WIN32)
#include <windows.h>
#include <conio.h>
#define snprintf sprintf_s 
#else
#include <sys/time.h>
#endif

#include "DynamsoftBarcodeReader.h"
#include "DynamsoftCommon.h"
using namespace dynamsoft::dbr;
typedef struct BarcodeFormatSet
{
	int barcodeFormatIds;
	int barcodeFormatIds_2;
}BarcodeFormatSet;

unsigned long GetTiming()
{
#if defined(_WIN64) || defined(_WIN32)
	return GetTickCount();
#else
	struct timeval timing;
	gettimeofday(&timing, NULL);
	return timing.tv_sec * 1000  + timing.tv_usec / 1000;
#endif
}

const int GetBarcodeFormatId(int iIndex, BarcodeFormatSet* barcodeFormatSet)
{
	int ret = 0;
	switch (iIndex)
	{
	case 1:
		barcodeFormatSet->barcodeFormatIds = BF_ALL;
		barcodeFormatSet->barcodeFormatIds_2 = BF2_POSTALCODE | BF2_DOTCODE;
		break;
	case 2:
		barcodeFormatSet->barcodeFormatIds = BF_ONED;
		barcodeFormatSet->barcodeFormatIds_2 = 0;
		break;
	case 3:
		barcodeFormatSet->barcodeFormatIds = BF_QR_CODE;
		barcodeFormatSet->barcodeFormatIds_2 = 0;
		break;
	case 4:
		barcodeFormatSet->barcodeFormatIds = BF_CODE_39;
		barcodeFormatSet->barcodeFormatIds_2 = 0;
		break;
	case 5:
		barcodeFormatSet->barcodeFormatIds = BF_CODE_128;
		barcodeFormatSet->barcodeFormatIds_2 = 0;
		break;
	case 6:
		barcodeFormatSet->barcodeFormatIds = BF_CODE_93;
		barcodeFormatSet->barcodeFormatIds_2 = 0;
		break;
	case 7:
		barcodeFormatSet->barcodeFormatIds = BF_CODABAR;
		barcodeFormatSet->barcodeFormatIds_2 = 0;
		break;
	case 8:
		barcodeFormatSet->barcodeFormatIds = BF_ITF;
		barcodeFormatSet->barcodeFormatIds_2 = 0;
		break;
	case 9:
		barcodeFormatSet->barcodeFormatIds = BF_INDUSTRIAL_25;
		barcodeFormatSet->barcodeFormatIds_2 = 0;
		break;
	case 10:
		barcodeFormatSet->barcodeFormatIds = BF_EAN_13;
		barcodeFormatSet->barcodeFormatIds_2 = 0;
		break;
	case 11:
		barcodeFormatSet->barcodeFormatIds = BF_EAN_8;
		barcodeFormatSet->barcodeFormatIds_2 = 0;
		break;
	case 12:
		barcodeFormatSet->barcodeFormatIds = BF_UPC_A;
		barcodeFormatSet->barcodeFormatIds_2 = 0;
		break;
	case 13:
		barcodeFormatSet->barcodeFormatIds = BF_UPC_E;
		barcodeFormatSet->barcodeFormatIds_2 = 0;
		break;
	case 14:
		barcodeFormatSet->barcodeFormatIds = BF_PDF417;
		barcodeFormatSet->barcodeFormatIds_2 = 0;
		break;
	case 15:
		barcodeFormatSet->barcodeFormatIds = BF_DATAMATRIX;
		barcodeFormatSet->barcodeFormatIds_2 = 0;
		break;
	case 16:
		barcodeFormatSet->barcodeFormatIds = BF_AZTEC;
		barcodeFormatSet->barcodeFormatIds_2 = 0;
		break;
	case 17:
		barcodeFormatSet->barcodeFormatIds = BF_CODE_39_EXTENDED;
		barcodeFormatSet->barcodeFormatIds_2 = 0;
		break;
	case 18:
		barcodeFormatSet->barcodeFormatIds = BF_MAXICODE;
		barcodeFormatSet->barcodeFormatIds_2 = 0;
		break;
	case 19:
		barcodeFormatSet->barcodeFormatIds = BF_GS1_DATABAR;
		barcodeFormatSet->barcodeFormatIds_2 = 0;
		break;
	case 20:
		barcodeFormatSet->barcodeFormatIds = BF_PATCHCODE;
		barcodeFormatSet->barcodeFormatIds_2 = 0;
		break;
	case 21:
		barcodeFormatSet->barcodeFormatIds = BF_GS1_COMPOSITE;
		barcodeFormatSet->barcodeFormatIds_2 = 0;
		break;
	case 22:
		barcodeFormatSet->barcodeFormatIds = 0;
		barcodeFormatSet->barcodeFormatIds_2 = BF2_POSTALCODE;
		break;
	case 23:
		barcodeFormatSet->barcodeFormatIds = 0;
		barcodeFormatSet->barcodeFormatIds_2 = BF2_DOTCODE;
		break;
	case 24:
		barcodeFormatSet->barcodeFormatIds = BF_MSI_CODE;
		barcodeFormatSet->barcodeFormatIds_2 = 0;
		break;
	default:
		ret = -1;
		break;
	}
	return ret;
}

void ToHexString(unsigned char* pSrc, int iLen, char* pDest)
{
	const char HEXCHARS[16] = { '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F' };

	int i;
	char* ptr = pDest;

	for (i = 0; i < iLen; ++i)
	{
		snprintf(ptr, 4, "%c%c ", HEXCHARS[(pSrc[i] & 0xF0) >> 4], HEXCHARS[(pSrc[i] & 0x0F) >> 0]);
		ptr += 3;
	}
}

bool GetImagePath(char* pImagePath)
{
	char pszBuffer[512] = { 0 };
	bool bExit = false;
	size_t iLen = 0;
	FILE* fp = NULL;
	while (1)
	{
		printf("\r\n>> Step 1: Input your image file's full path:\r\n");
		fgets(pszBuffer, 512, stdin);
		strtok(pszBuffer, "\n");

		iLen = strlen(pszBuffer);
		if (iLen > 0)
		{
			if (strlen(pszBuffer) == 1 && (pszBuffer[0] == 'q' || pszBuffer[0] == 'Q'))
			{
				bExit = true;
				break;
			}

			memset(pImagePath, 0, 512);
			if (pszBuffer[0] == '\'' || pszBuffer[0] == '\"' )
			{
				if (pszBuffer[iLen - 3] == '\'')
					memcpy(pImagePath, pszBuffer + 1, iLen - 4);
				else if (pszBuffer[iLen - 2] == '\'')
					memcpy(pImagePath, pszBuffer + 1, iLen - 3);
				else if(pszBuffer[iLen - 1] == '\"')
					memcpy(pImagePath, pszBuffer + 1, iLen - 2);
			}
			else
				memcpy(pImagePath, pszBuffer, iLen);

			fp = fopen(pImagePath, "rb");
			if (fp != NULL)
			{
				break;
			}
		}
		printf("Please input a valid path.\r\n");
	}
	return bExit;
}

bool SetBarcodeFormat(BarcodeFormatSet* iBarcodeFormatId)
{
	char pszBuffer[512] = { 0 };
	bool bExit = false;
	size_t iLen = 0;
	int iIndex = 0;
	while (1)
	{
		printf("\r\n>> Step 2: Choose a number for the format(s) of your barcode image:\r\n");
		printf("   1: All\r\n");
		printf("   2: OneD\r\n");
		printf("   3: QR Code\r\n");
		printf("   4: Code 39\r\n");
		printf("   5: Code 128\r\n");
		printf("   6: Code 93\r\n");
		printf("   7: Codabar\r\n");
		printf("   8: Interleaved 2 of 5\r\n");
		printf("   9: Industrial 2 of 5\r\n");
		printf("   10: EAN-13\r\n");
		printf("   11: EAN-8\r\n");
		printf("   12: UPC-A\r\n");
		printf("   13: UPC-E\r\n");
		printf("   14: PDF417\r\n");
		printf("   15: DATAMATRIX\r\n");
		printf("   16: AZTEC\r\n");
		printf("   17: Code 39 Extended\r\n");
		printf("   18: Maxicode\r\n");
		printf("   19: GS1 Databar\r\n");
		printf("   20: PatchCode\r\n");
		printf("   21: GS1 Composite\r\n");
		printf("   22: Postal  Code\r\n");
		printf("   23: DotCode\r\n");
		printf("   24: MSI Code\r\n");

		fgets(pszBuffer, 512, stdin);
		strtok(pszBuffer, "\n");

		iLen = strlen(pszBuffer);
		if (iLen > 0)
		{

			if (strlen(pszBuffer) == 1 && (pszBuffer[0] == 'q' || pszBuffer[0] == 'Q'))
			{
				bExit = true;
				break;
			}

			iIndex = atoi(pszBuffer);
			int ret = GetBarcodeFormatId(iIndex, iBarcodeFormatId);
			if (ret != -1)
				break;
		}

		if (bExit)
			break;

		printf("Please choose a valid number. \r\n");

	}
	return bExit;
}

void OutputResult(CBarcodeReader& reader, int errorcode, float time)
{
	char * pszTemp = NULL;
	char * pszTemp1 = NULL;
	char * pszTemp2 = NULL;
	int iRet = errorcode;
	pszTemp = (char*)malloc(4096);
	if (iRet != DBR_OK && iRet != DBRERR_MAXICODE_LICENSE_INVALID && iRet != DBRERR_AZTEC_LICENSE_INVALID && iRet != DBRERR_LICENSE_EXPIRED && iRet != DBRERR_QR_LICENSE_INVALID && iRet != DBRERR_GS1_COMPOSITE_LICENSE_INVALID &&
		iRet != DBRERR_1D_LICENSE_INVALID && iRet != DBRERR_PDF417_LICENSE_INVALID && iRet != DBRERR_DATAMATRIX_LICENSE_INVALID && iRet != DBRERR_GS1_DATABAR_LICENSE_INVALID && iRet != DBRERR_PATCHCODE_LICENSE_INVALID &&
		iRet != DBRERR_POSTALCODE_LICENSE_INVALID && iRet != DBRERR_DOTCODE_LICENSE_INVALID && iRet != DBRERR_DPM_LICENSE_INVALID && iRet != DBRERR_IRT_LICENSE_INVALID && iRet != DMERR_NO_LICENSE && iRet != DMERR_TRIAL_LICENSE)
	{
		snprintf(pszTemp, 4096, "Failed to read barcode: %s\r\n", CBarcodeReader::GetErrorString(iRet));
		printf("%s", pszTemp);
		free(pszTemp);
		return;
	}

	TextResultArray *paryResult = NULL;
	reader.GetAllTextResults(&paryResult);

	if (paryResult->resultsCount == 0)
	{
		snprintf(pszTemp, 4096, "No barcode found. Total time spent: %.3f seconds.\r\n", time);
		printf("%s", pszTemp);
		free(pszTemp);
		CBarcodeReader::FreeTextResults(&paryResult);
		return;
	}

	snprintf(pszTemp, 4096, "Total barcode(s) found: %d. Total time spent: %.3f seconds\r\n\r\n", paryResult->resultsCount, time);
	printf("%s", pszTemp);
	for (int iIndex = 0; iIndex < paryResult->resultsCount; iIndex++)
	{
		snprintf(pszTemp, 4096, "Barcode %d:\r\n", iIndex + 1);
		printf("%s", pszTemp);
		if (paryResult->results[iIndex]->barcodeFormat != 0)
		{
			snprintf(pszTemp, 4096, "    Type: %s\r\n", paryResult->results[iIndex]->barcodeFormatString);
		}
		else
		{
			snprintf(pszTemp, 4096, "    Type: %s\r\n", paryResult->results[iIndex]->barcodeFormatString_2);
		}

		printf("%s", pszTemp);
		snprintf(pszTemp, 4096, "    Value: %s\r\n", paryResult->results[iIndex]->barcodeText);
		printf("%s", pszTemp);

		pszTemp1 = (char*)malloc(paryResult->results[iIndex]->barcodeBytesLength * 3 + 1);
		pszTemp2 = (char*)malloc(paryResult->results[iIndex]->barcodeBytesLength * 3 + 100);
		ToHexString(paryResult->results[iIndex]->barcodeBytes, paryResult->results[iIndex]->barcodeBytesLength, pszTemp1);
		snprintf(pszTemp2, paryResult->results[iIndex]->barcodeBytesLength * 3 + 100, "    Hex Data: %s\r\n", pszTemp1);
		printf("%s", pszTemp2);
		free(pszTemp1);
		free(pszTemp2);
	}

	free(pszTemp);
	CBarcodeReader::FreeTextResults(&paryResult);
}

int main(int argc, const char* argv[])
{
	const char* pszTemplateName = NULL;
	BarcodeFormatSet iBarcodeFormatId = { 0,0 };
	char pszBuffer[512] = { 0 };
	char pszImageFile[512] = { 0 };
	int iIndex = 0;
	int iRet = -1;
	char * pszTemp = NULL;
	char * pszTemp1 = NULL;
	unsigned long ullTimeBegin = 0;
	unsigned long ullTimeEnd = 0;

	size_t iLen = 0;
	FILE* fp = NULL;
	bool bExit = false;
	char szErrorMsg[256];
	PublicRuntimeSettings runtimeSettings;

	printf("*************************************************\r\n");
	printf("Welcome to Dynamsoft Barcode Reader Demo\r\n");
	printf("*************************************************\r\n");
	printf("Hints: Please input 'Q' or 'q' to quit the application.\r\n");

	CBarcodeReader reader;
	reader.InitLicense("t0068MgAAAC4m3vBqmGZTtroTtWT8STe3QFyOlwiLqwtB7ghY+8pU01rgU3JedgtcsxyRHnys1N3qjCMZ2AprRCavlIpLbAk=");


	while (1)
	{
		bExit = GetImagePath(pszImageFile);
		if (bExit)
			break;

		bExit = SetBarcodeFormat(&iBarcodeFormatId);
		if (bExit)
			break;

		//Best coverage settings
		reader.InitRuntimeSettingsWithString("{\"ImageParameter\":{\"Name\":\"BestCoverage\",\"DeblurLevel\":9,\"ExpectedBarcodesCount\":512,\"ScaleDownThreshold\":100000,\"LocalizationModes\":[{\"Mode\":\"LM_CONNECTED_BLOCKS\"},{\"Mode\":\"LM_SCAN_DIRECTLY\"},{\"Mode\":\"LM_STATISTICS\"},{\"Mode\":\"LM_LINES\"},{\"Mode\":\"LM_STATISTICS_MARKS\"}],\"GrayscaleTransformationModes\":[{\"Mode\":\"GTM_ORIGINAL\"},{\"Mode\":\"GTM_INVERTED\"}]}}", CM_OVERWRITE, szErrorMsg, 256);
		//Best speed settings
		//reader.InitRuntimeSettingsWithString("{\"ImageParameter\":{\"Name\":\"BestSpeed\",\"DeblurLevel\":3,\"ExpectedBarcodesCount\":512,\"LocalizationModes\":[{\"Mode\":\"LM_SCAN_DIRECTLY\"}],\"TextFilterModes\":[{\"MinImageDimension\":262144,\"Mode\":\"TFM_GENERAL_CONTOUR\"}]}}",CM_OVERWRITE,szErrorMsg,256);
		//Balance settings
		//reader.InitRuntimeSettingsWithString("{\"ImageParameter\":{\"Name\":\"Balance\",\"DeblurLevel\":5,\"ExpectedBarcodesCount\":512,\"LocalizationModes\":[{\"Mode\":\"LM_CONNECTED_BLOCKS\"},{\"Mode\":\"LM_STATISTICS\"}]}}",CM_OVERWRITE,szErrorMsg,256);

		reader.GetRuntimeSettings(&runtimeSettings);
		runtimeSettings.barcodeFormatIds = iBarcodeFormatId.barcodeFormatIds;
		runtimeSettings.barcodeFormatIds_2 = iBarcodeFormatId.barcodeFormatIds_2;
		iRet = reader.UpdateRuntimeSettings(&runtimeSettings, szErrorMsg, 256);
		if (iRet != DBR_OK)
		{
			printf("Error code: %d. Error message: %s\n", iRet, szErrorMsg);
			return -1;
		}

		ullTimeBegin = GetTiming();
		iRet = reader.DecodeFile(pszImageFile, pszTemplateName);
		ullTimeEnd = GetTiming();
		OutputResult(reader, iRet, (((float)(ullTimeEnd - ullTimeBegin)) / 1000));

	}

	return 0;
}

File: samples/Java/BarcodeReaderDemo/src/com/dynamsoft/demo/BarcodeReaderDemo.java
package com.dynamsoft.demo;

import java.util.Date;
import java.io.*;

import com.dynamsoft.barcode.Barcode;
import com.dynamsoft.barcode.BarcodeReader;
import com.dynamsoft.barcode.ReadResult;

public final class BarcodeReaderDemo {
	private BarcodeReaderDemo() {
	}
	
	private static long GetFormat(int iIndex) {
		long lFormat = 0;

		switch (iIndex) {
		case 1:
			lFormat = Barcode.OneD | Barcode.QR_CODE
					| Barcode.PDF417 | Barcode.DATAMATRIX;
			break;
		case 2:
			lFormat = Barcode.OneD;
			break;
		case 3:
			lFormat = Barcode.QR_CODE;
			break;
		case 4:
			lFormat = Barcode.CODE_39;
			break;
		case 5:
			lFormat = Barcode.CODE_128;
			break;
		case 6:
			lFormat = Barcode.CODE_93;
			break;
		case 7:
			lFormat = Barcode.CODABAR;
			break;
		case 8:
			lFormat = Barcode.ITF;
			break;
		case 9:
			lFormat = Barcode.INDUSTRIAL_25;
			break;
		case 10:
			lFormat = Barcode.EAN_13;
			break;
		case 11:
			lFormat = Barcode.EAN_8;
			break;
		case 12:
			lFormat = Barcode.UPC_A;
			break;
		case 13:
			lFormat = Barcode.UPC_E;
			break;
		case 14:
			lFormat = Barcode.PDF417;
			break;
		case 15:
			lFormat = Barcode.DATAMATRIX;
			break;
		default:
			break;
		}

		return lFormat;
	}
	
	public static void main(String[] args) throws Exception {

		int iMaxCount = 0;
		long lFormat = -1;
		int iIndex = 0;
		String pszImageFile = null;
		String strLine;
		boolean bExitFlag = false;

		System.out.println("*************************************************");
		System.out.println("Welcome to Dynamsoft Barcode Reader Demo");
		System.out.println("*************************************************");
		System.out
				.println("Hints: Please input 'Q'or 'q' to quit the application.");

		BufferedReader cin = new BufferedReader(new java.io.InputStreamReader(
				System.in));

		while (true) {
			iMaxCount = 0x7FFFFFFF;
			lFormat = (Barcode.OneD | Barcode.QR_CODE
					| Barcode.PDF417 | Barcode.DATAMATRIX);

			while (true) {
				System.out.println();
				System.out
						.println(">> Step 1: Input your image file's full path:");
				strLine = cin.readLine();
				if (strLine != null && strLine.trim().length() > 0) {
					strLine = strLine.trim();
					if (strLine.equalsIgnoreCase("q")){
						bExitFlag = true;
						break;
					}
						
					if (strLine.length() >= 2 && strLine.charAt(0) == '\"'
							&& strLine.charAt(strLine.length() - 1) == '\"') {
						pszImageFile = strLine.substring(1, strLine.length() - 1);
					} else {
						pszImageFile = strLine;
					}
					
					java.io.File file = new java.io.File(pszImageFile);
					if(file.exists() && file.isFile())
						break;
				}
				
				System.out.println("Please input a valid path.");
			}
			
			if(bExitFlag)
				break;

			while (true) {
				System.out.println();
				System.out
						.println(">> Step 2: Choose a number for the format(s) of your barcode image:");
				System.out.println("   1: All");
				System.out.println("   2: OneD");
				System.out.println("   3: QR Code");
				System.out.println("   4: Code 39");
				System.out.println("   5: Code 128");
				System.out.println("   6: Code 93");
				System.out.println("   7: Codabar");
				System.out.println("   8: Interleaved 2 of 5");
				System.out.println("   9: Industrial 2 of 5");
				System.out.println("   10: EAN-13");
				System.out.println("   11: EAN-8");
				System.out.println("   12: UPC-A");
				System.out.println("   13: UPC-E");
				System.out.println("   14: PDF417");
				System.out.println("   15: DATAMATRIX");

				strLine = cin.readLine();
				if (strLine.length() > 0) {
					try {
						iIndex = Integer.parseInt(strLine);
						lFormat = GetFormat(iIndex);
						if (lFormat != 0)
							break;
					} catch (Exception exp) {
					}
				}

				System.out.println("Please choose a valid number. ");

			}

			while (true) {
				System.out.println();
				System.out
						.println(">> Step 3: Input the maximum number of barcodes to read per page: ");

				strLine = cin.readLine();
				if (strLine.length() > 0) {
					try {
						iMaxCount = Integer.parseInt(strLine);
						if (iMaxCount > 0)
							break;
					} catch (Exception exp) {
					}
				}

				System.out.println("Please re-input the correct number again.");
			}

			System.out.println();
			System.out.println("Barcode Results:");
			System.out.println("----------------------------------------------------------");

			// Set license
			BarcodeReader br = new BarcodeReader("t0068MgAAABs0soPfOcktn1WIaQwU5tPkLklx8PbtKusKGedkkCTDIQldAxDlOkitjsoOoUHYq9Zxro5YEVTQ7/oqoIcoGzQ=");
			
			// Read barcode
			long ullTimeBegin = new Date().getTime();			
			ReadResult result = br.readFile(pszImageFile, lFormat, iMaxCount);			
			long ullTimeEnd = new Date().getTime();
			
			if (result.errorCode != BarcodeReader.DBR_OK && result.errorCode != BarcodeReader.DBRERR_LICENSE_EXPIRED && result.errorCode != BarcodeReader.DBRERR_LICENSE_INVALID) {
				System.out.println(result.errorString);
				continue;
			}

			// Output barcode result
			/*
			 * Total barcode(s) found: 2. Total time spent: 0.218 seconds.
			 * 
			 * Barcode 1: Page: 1 Type: CODE_128 Value: Zt=-mL-94 Region: {Left:
			 * 100, Top: 20, Width: 100, Height: 40}
			 * 
			 * Barcode 2: Page: 1 Type: CODE_39 Value: Dynamsoft Region: {Left:
			 * 100, Top: 200, Width: 180, Height: 30}
			 */

			String pszTemp;				

			if (result.barcodes == null || result.barcodes.length == 0) {
				pszTemp = String.format(
						"No barcode found. Total time spent: %.3f seconds.", ((float) (ullTimeEnd - ullTimeBegin) / 1000));
				System.out.println(pszTemp);
			} else {
				pszTemp = String.format("Total barcode(s) found: %d. Total time spent: %.3f seconds.", result.barcodes.length, ((float) (ullTimeEnd - ullTimeBegin) / 1000));
				System.out.println(pszTemp);			

				for (iIndex = 0; iIndex < result.barcodes.length; iIndex++) {
					Barcode barcode = result.barcodes[iIndex];
					
					pszTemp = String.format("  Barcode %d:", iIndex + 1);
					System.out.println(pszTemp);
					pszTemp = String.format("    Page: %d", barcode.pageNumber);
					System.out.println(pszTemp);
					pszTemp = String.format("    Type: %s", barcode.formatString);						
					System.out.println(pszTemp);				
					pszTemp = "    Value: " + barcode.displayValue;
					System.out.println(pszTemp);		
					
					
					pszTemp = String.format("    Region: {Left: %d, Top: %d, Width: %d, Height: %d}",
									barcode.boundingBox.x, barcode.boundingBox.y, barcode.boundingBox.width, barcode.boundingBox.height);
	
					System.out.println(pszTemp);
					System.out.println();
				}
			}
		}
	}
}
